gcsio/src/main/java/com/google/cloud/hadoop/gcsio/MetricsPublisher.java

```diff
@@ -0,0 +1,243 @@
+   * @param n - Value of the measurement recorded
+   */
+  void recordLong(TagKey[] keys, String[] values, MeasureLong ml, Long n);
+}
+
+/** Publishes open-census measurements to StackDriver */
+class StackDriverMetricsPublisher implements MetricsPublisher {
+
+  @VisibleForTesting static final Object monitor = new Object();
+
+  @GuardedBy("monitor")
+  private static boolean initialized = false;
+
+  private static final TagMetadata TAG_METADATA_NO_PROPAGATION =
+      TagMetadata.create(TagTtl.NO_PROPAGATION);
+
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  static final MeasureLong LATENCY_MS =
+      MeasureLong.create("gcsio/latency", "The latency in milliseconds ", "ms");
+
+  static final MeasureLong MESSAGE_LATENCY_MS =
+      MeasureLong.create(
+          "gcsio/message/latency", "The latency in milliseconds per gcs message loop", "ms");
+
+  static final MeasureLong REQUESTS =
+      MeasureLong.create(
+          "gcsio/requests", "The distribution of retry attempts for gcs calls", "By");
+
+  static final MeasureLong REQUEST_RETRIES =
+      MeasureLong.create("gcsio/retries", "The distribution of retry attempts for gcs calls", "By");
+
+  static final TagKey KEY_METHOD = TagKey.create("method");
+  static final TagKey KEY_STATUS = TagKey.create("status");
+  static final TagKey KEY_ERROR = TagKey.create("error");
+  static final TagKey KEY_THREAD = TagKey.create("thread");
+  static final TagKey KEY_PROTOCOL = TagKey.create("protocol");
+
+  static final List<TagKey> TAG_KEYS =
+      Collections.unmodifiableList(
+          Arrays.asList(KEY_METHOD, KEY_STATUS, KEY_ERROR, KEY_THREAD, KEY_PROTOCOL));
+
+  private static final Tagger tagger = Tags.getTagger();
+  private static final StatsRecorder statsRecorder = Stats.getStatsRecorder();
+
+  public static MetricsPublisher create(String projectId, Credential credential) {
+    CredentialAdapter credentialAdapter = new CredentialAdapter(credential);
+    return create(projectId, credentialAdapter);
+  }
+
+  public static MetricsPublisher create(String projectId, Credentials credentials) {
+    try {
+      registerAllViews();
+      setupStackDriverExporter(projectId, credentials);
+    } catch (Exception e) {
+      logger.atWarning().withCause(e).log("Exception while registering metrics publisher");
+      return new NoOpMetricsPublisher();
+    }
+    return new StackDriverMetricsPublisher();
+  }
+
+  private static void setupStackDriverExporter(String projectId, Credentials credentials)
+      throws IOException {
+    synchronized (monitor) {
+      if (!initialized) {
+        StackdriverStatsExporter.createAndRegister(
+            StackdriverStatsConfiguration.builder()
+                .setCredentials(credentials)
+                .setProjectId(projectId)
+                .setExportInterval(fromMillis(60000))
+                .build());
+        Runtime.getRuntime().addShutdownHook(new Thread(StackdriverStatsExporter::unregister));
```

Comment:
Is this really necessary?

Reply:
`unregister` method flushes the metrics. It is necessary for the metrics to not get lost. Within spark and connector context, a file system can get closed on a single jvm multiple times. So, we don't have a clean handle to flush and close the metrics publisher